static/js/recent_topics.js

```diff
@@ -0,0 +1,55 @@
+const topics = new Map(); // Key is stream-id:topic.
+
+exports.process_messages = function (messages) {
+    messages.forEach(exports.process_message);
+};
+
+function reduce_message(msg) {
+    return {
+        id: msg.id,
+        timestamp: msg.timestamp,
+        stream_id: msg.stream_id,
+        stream_name: msg.stream,
+        topic: msg.topic,
+        sender_id: msg.sender_id,
+        type: msg.type,
+    };
+}
```

Comment:
Is this `reduce_message` system, rather than pointing to the message from `message_store` (without copying), useful?

Reply:
I wanted to store some custom fields per message for which I think this can be useful. (I was tracking more messages initially too.) Let me see if this is required now or not.

Reply:
I think it removes a lot of code clutter to have some msg properties stored in here. So, we should use it.

Reply:
What properties do you store here that aren't properties of our normal message objects in `message_store`?

Right now, we have the nice arrangement in our codebase that there's exactly one type for a Message object in the frontend codebase, stored in a shared cache and usable everywhere that message ID is accessed.

I'd like to preserve that model unless there's a really compelling reason not to do so, since it makes the overall codebase easier to work with.  If we need to add a couple additional computed values in `message_store`, we can do that.

Reply:
Okay then, I agree that we should maintain the arrangement you mentioned. I will update this.